src/frontend/src/optimizer/share_source_rewriter.rs

```diff
@@ -0,0 +1,116 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{HashMap, HashSet};
+
+use itertools::Itertools;
+
+use crate::catalog::SourceId;
+use crate::optimizer::plan_node::{
+    LogicalShare, LogicalSource, PlanNodeId, PlanTreeNode, StreamShare,
+};
+use crate::optimizer::plan_rewriter::PlanRewriter;
+use crate::optimizer::PlanVisitor;
+use crate::PlanRef;
+
+#[derive(Debug, Clone, Default)]
+pub struct ShareSourceRewriter {
+    // Source id need to be shared.
```

Comment:
```suggestion
    /// Source id need to be shared.
```

Ditto. for other occurrences.